src/preview.ts

```diff
@@ -188,12 +188,21 @@ export class PreviewServer {
           if (!isEnoent(error)) throw error; // internal error
         }
 
-        // If this path ends with .html, then redirect to drop the .html. TODO:
-        // Check for the existence of the .md file first.
+        // If this path ends with .html and cleanUrls are enabled, then redirect
+        // to drop the .html.
         if (extname(path) === ".html") {
-          res.writeHead(302, {Location: join(dirname(pathname), basename(pathname, ".html")) + url.search});
-          res.end();
-          return;
+          path = path.slice(0, -".html".length);
+          pathname = pathname.slice(0, -".html".length);
+          if (config.md.normalizeLink("./hello") === "./hello") {
```

Comment:
This seems a little icky. Wouldn’t it be better to check the value of `config.md.normalizeLink(path)` instead, and then redirect to the normalized value? Otherwise we’re essentially duplicating the logic of `normalizeLink` and assuming its behavior. Or if we want to assume its behavior, then we should expose `config.cleanUrls` instead of trying to invert it by calling `md.normalizeLink`.